server/src/test/java/org/elasticsearch/action/ingest/WriteableIngestDocumentTests.java

```diff
@@ -147,4 +151,71 @@ public void testToXContent() throws IOException {
         IngestDocument serializedIngestDocument = new IngestDocument(toXContentSource, toXContentIngestMetadata);
         assertThat(serializedIngestDocument, equalTo(serializedIngestDocument));
     }
+
+    public void testSourceFromXContentWithByteArray() throws IOException {
+        WriteableIngestDocument testInstance = new WriteableIngestDocument(RandomDocumentPicks.randomIngestDocument(random(), true));
+        XContentType xContentType = randomFrom(XContentType.values());
+        BytesReference shuffled = toShuffledXContent(testInstance, xContentType, ToXContent.EMPTY_PARAMS, false);
+        XContentParser parser = createParser(XContentFactory.xContent(xContentType), shuffled);
+        WriteableIngestDocument parsed = doParseInstance(parser);
+        assertTrue(
+            testSourceMapEquality(
+                testInstance.getIngestDocument().getSourceAndMetadata(),
+                parsed.getIngestDocument().getSourceAndMetadata()
+            )
+        );
+    }
+
+    private boolean testSourceMapEquality(Map<String, Object> first, Map<String, Object> second) {
+        if (first == null) {
+            return second == null;
+        } else if (second == null) {
+            return false;
+        } else if (first.size() != second.size()) {
+            return false;
+        } else {
+            for (Map.Entry<String, Object> entry: first.entrySet()) {
+                Object value = entry.getValue();
+                Object value2;
+                if ((value2 = second.get(entry.getKey())) != null) {
+                    if (value2.getClass() == value.getClass()) {
+                        if (value instanceof byte[]) {
+                            if (!Arrays.equals((byte[])value, (byte[])value2)) {
+                                return false;
+                            }
+                        } else if (value instanceof Map) {
+                            //noinspection unchecked
+                            if (!testSourceMapEquality((Map<String, Object>)value, (Map<String, Object>)value2)) {
+                                return false;
+                            }
+                        } else if (!value.equals(value2)) {
+                            return false;
+                        }
+                    }
+                } else {
+                    return false;
+                }
+            }
+            return true;
+        }
+    }
+
+    @Override
+    protected WriteableIngestDocument createTestInstance() {
+        // We want to create a test instance without byte arrays. For testing byte arrays see testSourceFromXContentWithByteArray.
+        // This is needed because for comparing byte[] Object.equal fails
```

Comment:
just an idea: could we maybe try to leave byte arrays but modify the equality assertion overriding assertEqualInstances ?

Reply:
I don't know. `equals` and `hashCode` are already implemented for this so I did it like this. If you think it is better to override `assertEqualInstances` I will make the change.

Reply:
maybe give it a try and see if it gets better. I think it may, but I would have to try as well to see what it looks like :)